package-testing/android-relay/app/src/main/java/cloud/eppo/android/sdkrelay/TestClientActivity.kt

```diff
@@ -0,0 +1,279 @@
+            " assigned to " +
+            assignment.experiment)
+    Log.d(TAG, msg)
+    appendAssignmentLog(msg)
+  }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+
+    // Initialize the socket connection
+    SocketHandler.setSocket(
+        host = BuildConfig.TEST_RUNNER_HOST, port = BuildConfig.TEST_RUNNER_PORT)
+
+    // Set listeners
+    SocketHandler.mSocket.on("connect_error") { args -> handleConnectError(args) }
+    SocketHandler.mSocket.on("connect") { args -> handleConnect(args) }
+    SocketHandler.mSocket.on("disconnect") { args -> handleDisconnect(args) }
+    SocketHandler.mSocket.on("/sdk/reset") { args -> handleReset(args) }
+    SocketHandler.mSocket.on("/flags/v1/assignment") { args -> handleAssignment(args) }
+
+    // Set the UI
+    setContent {
+      EppoApplicationTheme {
+        Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
+          TestClientScreen(status, assignmentLog, Modifier.padding(innerPadding))
+        }
+      }
+    }
+
+    // Connect to the test runner
+    status.postValue("Connecting")
+    reInitializeEppoClient().thenRun({ SocketHandler.establishConnection() })
+  }
+
+  private fun handleConnect(args: Array<Any>?) {
+    Log.d(TAG, "Connected")
+    status.postValue("Connected")
+    sendReady()
+  }
+
+  private fun handleConnectError(args: Array<Any>) {
+    val exception: EngineIOException = args[0] as EngineIOException
+    status.postValue("Connection Error")
+    Log.e(TAG, "Connection error", exception)
+  }
+
+  private fun handleAssignment(args: Array<Any>) {
+    Log.d(TAG, "Assignment Requested")
+    status.postValue("Assigning")
+
+    val requestObj = (args[0] as JSONObject)
+    val assignmentRequest =
+        AssignmentRequest(
+            flag = requestObj.getString("flag"),
+            subjectKey = requestObj.getString("subjectKey"),
+            assignmentType = requestObj.getString("assignmentType"),
+            subjectAttributes = requestObj.getJSONObject("subjectAttributes"),
+            defaultValue = requestObj.get("defaultValue"))
+
+    // Ack function for responding to the server.
+    val ack = args.last() as Ack
+
+    val client = EppoClient.getInstance()
+
+    val eppoValues: MutableMap<String, EppoValue> = mutableMapOf()
+    assignmentRequest.subjectAttributes.keys().forEach {
+      val value = assignmentRequest.subjectAttributes.get(it.toString())
+      val typedValue: EppoValue =
+          when (value) {
+            is Int -> EppoValue.valueOf(value.toDouble())
+            is String -> EppoValue.valueOf(value)
+            is Double -> EppoValue.valueOf(value)
+            is Boolean -> EppoValue.valueOf(value)
+            is List<*> -> EppoValue.valueOf(value.map { insideIt -> insideIt.toString() })
+            else -> {
+              Log.e(TAG, "Invalid or null subject attribute type.")
+              EppoValue.nullValue()
+            }
+          }
```

Comment:
😮

Reply:
ya. it's pretty slick eh?

Reply:
nit: you could use `assignmentRequest.subjectAttributes.map {`

Reply:
`org.json.JSONObject` doesn't seem to have a `map` function. Per its docs, it has an entrySet method in java, but that doesn't appear to be a thing in Kotlin. Je suis le confused.